### PR TITLE
Feature/add drug exposure table

### PR DESCRIPTION
**Pull request by calmacx**
For processing ALSPAC with the latest set of rules, we need to add the drug exposure table to the pythonic version of the CDM in co-connect-tools

### CODE REVIEW

File: coconnect/cdm/objects/common.py [hunk 1]
```diff
@@ -27,6 +27,7 @@ def __init__(self):
         super().__init__()
         self['INTEGER'] = lambda x : pd.to_numeric(x,errors='coerce').astype('Int64')
         self['FLOAT'] = lambda x : pd.to_numeric(x,errors='coerce').astype('Float64')
+        self['VARCHAR(MAX)'] = lambda x : x.fillna('').astype(str).apply(lambda x: x)
```
Comment (calmacx) [inline coconnect/cdm/objects/common.py:30]: see #66

File: coconnect/cdm/objects/common.py [hunk 2]
```diff
@@ -80,11 +81,17 @@ def get_field_names(self):
...
             field
             for field in self.fields
             if getattr(self,field).pk == True
         ]
+
+        if len(retval) == 0:
+            #warning, no pk has been set on any field
+            retval = self.fields[0]
+        
+        return retval
```
Comment (calmacx) [inline coconnect/cdm/objects/common.py:94]: minor fix/protection, could be better suited in a separate PR

File: coconnect/cdm/objects/common.py [hunk 3]
```diff
@@ -237,7 +244,10 @@ def finalise(self,df):
             
             ndiff = nbefore - nafter
             if ndiff>0:
-                self.logger.warning(f"Requiring non-null values in {field} removed {ndiff} rows, leaving {nafter} rows.")
+                log = self.logger.warning
+                if nafter == 0:
+                    log = self.logger.error
+                log(f"Requiring non-null values in {field} removed {ndiff} rows, leaving {nafter} rows.")
```
Comment (calmacx) [inline coconnect/cdm/objects/common.py:250]: change the logging level to an error if requiring non-null values of a given field removes everything

Comment (spco) [reply coconnect/cdm/objects/common.py:250]: Could you one-line to make the logic a little clearer, like `log = self.logger.warning if nafter > 0 else self.logger.error`. That makes it perhaps clearer that it's dependent on `nafter` rather than this "set default then modify" pattern.

File: coconnect/cdm/objects/common.py (hunk 1 above)
Comment (spco) [reply coconnect/cdm/objects/common.py:30]: I have messaged Tom re this issue - will let you know if I hear anything. Do you consider this ready to merge as is, or would prefer to defer until `MAX` is understood?

File: coconnect/cdm/objects/common.py (hunk 3 above)
Comment (calmacx) [reply coconnect/cdm/objects/common.py:249]: yeh nice, implemented that now

File: coconnect/cdm/objects/common.py (hunk 1 above)
Comment (calmacx) [reply coconnect/cdm/objects/common.py:30]: Thanks! I think we're good to merge. 

I've never been able to test BCLink myself, as you know, so not even 100% sure this stuff is necessary or even up-to-date. It's probably highly unlikely this ever causes us an issue.

If there are BCLink upload issues, this formatting part of the code would be one of the first things I/we could check.

**Review by spco — APPROVED** [review]
